src/java/org/apache/cassandra/db/compaction/CompactionTask.java

```diff
@@ -256,27 +260,27 @@ public boolean apply(SSTableReader sstable)
                 long lastBytesScanned = 0;
 
                 activeCompactions.beginCompaction(ci);
-                try (CompactionAwareWriter writer = getCompactionAwareWriter(cfs, getDirectories(), transaction, actuallyCompact))
+                try (AutoCloseable resource = getCompactionAwareWriter(actuallyCompact, ci))
                 {
                     // Note that we need to re-check this flag after calling beginCompaction above to avoid a window
                     // where the compaction does not exist in activeCompactions but the CSM gets paused.
                     // We already have the sstables marked compacting here so CompactionManager#waitForCessation will
                     // block until the below exception is thrown and the transaction is cancelled.
                     if (!controller.cfs.getCompactionStrategyManager().isActive())
                         throw new CompactionInterruptedException(ci.getCompactionInfo());
-                    estimatedKeys = writer.estimatedKeys();
-                    while (ci.hasNext())
+                    estimatedKeys = ci.estimatedKeys();
+                    while (ci.processNextPartitionKey())
                     {
-                        if (writer.append(ci.next()))
-                            totalKeysWritten++;
-
-                        ci.setTargetDirectory(writer.getSStableDirectory().path());
-                        long bytesScanned = scanners.getTotalBytesScanned();
+                        long bytesScanned = ci.getTotalBytesScanned();
 
-                        // Rate limit the scanners, and account for compression
-                        CompactionManager.instance.compactionRateLimiterAcquire(limiter, bytesScanned, lastBytesScanned, compressionRatio);
+                        // If we ingested less than a MB, keep going
+                        if (bytesScanned - lastBytesScanned > MEGABYTE)
+                        {
+                            // Rate limit the scanners, and account for compression
+                            CompactionManager.instance.compactionRateLimiterAcquire(limiter, bytesScanned, lastBytesScanned, compressionRatio);
```

Comment:
should we report the remaining part once the loop is completed to reflect accurately all scanned bytes in org.apache.cassandra.metrics.CompactionMetrics#bytesCompactedThroughput ?